feat: added new evaluate command to check the score of the asyncapi document

AayushSaini101 opened the pull request:
Resolves: #1131

* Started implementing new command to evaluate the asyncapi file:
* Use four parameters to describe the score of file:
 
```
 if (document?.info().hasDescription()) {
      scoreEvalate+=0.15;
    }
    if (document?.info().hasLicense()) {
      scoreEvalate+=0.25;
    }
    if (!document?.servers().isEmpty()) {
      scoreEvalate+=0.25;
    }
    if (!document?.channels().isEmpty()) {
      scoreEvalate+=0.35;
    }
    const finalScore = (scoreEvalate/1)*100;
```

Shurtu-gal commented:
@AayushSaini101 do we need a new command for the linked issue ?

AayushSaini101 commented:
> #1131

I guess it would be if we add new command for this, What do you think ? @Amzani @peter-rr @Souvikns

Shurtu-gal commented:
> > #1131
> 
> I guess it would be if we add new command for this, What do you think ? @Amzani @peter-rr @Souvikns

Was seeing the incorrect issue then 👍🏻

Amzani commented on src/commands/evaluate.ts:
```diff
@@ -0,0 +1,60 @@
+    }
+    const finalScore = (scoreEvaluate/1)*100;
+
+    this.log('The score of the asyncapi document is ', +finalScore);
```
@AayushSaini101 check if we can use the new UI library for this (see https://github.com/asyncapi/cli/issues/1214)

Amzani commented:
> @AayushSaini101 do we need a new command for the linked issue ?

This might be used under `validate` command, however, for a better DX and maintainability having a dedicated command looks like a better approach.

Shurtu-gal commented:
> > @AayushSaini101 do we need a new command for the linked issue ?
> 
> This might be used under `validate` command, however, for a better DX and maintainability having a dedicated command looks like a better approach.

Agreed, how about adding it as a flag `--evaluate`. Any way parsing will be done for both validate and evaluate. Could think some more ideas 🤔.

peter-rr commented:
IMO, I would opt for just adding a flag to `validate` command since the feature to be added should work like an extension for the validation process. Also I think `--score` would be semantically more clear and easier to understand than `--evaluate`. So the resulting command could be something like `asyncapi validate --score`.

WDYT? @AayushSaini101 @Amzani @Shurtu-gal

AayushSaini101 commented on src/commands/evaluate.ts:
```diff
@@ -0,0 +1,60 @@
+    if (!document?.channels().isEmpty()) {
+      scoreEvaluate+=0.35;
+    }
+    const finalScore = (scoreEvaluate/1)*100;
```
@peter-rr In the initial stage, the total weight of the coefficient is 1 that's why we need to divide the sum by 1, to get 100 Score for a valid asyncapi document.

Amzani commented:
> IMO, I would opt for just adding a flag to validate command since the feature to be added should work like an extension for the validation process. Also I think --score would be semantically more clear and easier to understand than --evaluate. So the resulting command could be something like asyncapi validate --score.

Evaluation and Validation are closely related, having `asyncapi validate --score` looks good to me as long as we can make the new feature easily maintainable, and users can discover it.

Shurtu-gal commented:
/update

There are some issues with sonar-cloud as well @AayushSaini101.

AayushSaini101 commented:
@Shurtu-gal @Amzani @Souvikns I have added score flag in the existing validate command, and added a test case to verify the score and command. Ready for review. : ) Thanks

Amzani commented on src/commands/validate.ts:
```diff
@@ -35,4 +61,26 @@ export default class Validate extends Command {
     }
   }
+
+  private async calculateScore(filePath:string) {
```
I suggest to extract the score logic and move it under `models` or `utils` for now. It will make this https://github.com/asyncapi/cli/pull/1200 easy.

AayushSaini101 replied:
Moved to Utils Thanks

AayushSaini101 commented on src/commands/validate.ts:
```diff
@@ -4,6 +4,24 @@ import { validate, validationFlags, ValidateOptions, ValidationStatus } from '..
 import { load } from '../models/SpecificationFile';
 import { specWatcher } from '../globals';
 import { watchFlag } from '../flags';
+import { Parser} from '@asyncapi/parser/cjs';
```
Thanks, I missed that Done @Amzani

AayushSaini101 commented on src/utils/parsingFile.ts:
```diff
@@ -0,0 +1,49 @@
+import { load } from 'js-yaml';
```
Reverted thanks

Shurtu-gal commented on src/utils/scoreCalculator.ts:
```diff
@@ -0,0 +1,20 @@
+  }
+  const finalScore = (scoreEvaluate / 1) * 100;
+  
+  return `The score of the asyncapi document is ${finalScore}`;
```
How about returning the score instead of a log string. As this makes it easier to use later in another scenarios if we want.

AayushSaini101 replied:
@Shurtu-gal In my opinion, proper logging would be better that enhanve the developer experience as well. @Amzani What do you think on this ?

Shurtu-gal replied:
Yeah, I am not suggesting to take out the log message, just return the score from this function. And put logging where this function is consumed.

This way this function can be reused if needed.

Shurtu-gal reviewed (approved):
LGTM from my side. 🚀

Let's wait for @Amzani as well.

Amzani commented on src/commands/validate.ts:
```diff
@@ -12,17 +13,27 @@ export default class Validate extends Command {
     watch: watchFlag(),
     ...validationFlags(),
+    score: Flags.boolean({
+      description: 'Compute the score of the AsyncAPI document',
```
Users might might not understand how score is calculated, could we add more description and details about that ?

AayushSaini101 replied:
Done @Amzani

Amzani reviewed (changes requested):
Great PR! only thing missing for me is describing how score is calculated for users.

AayushSaini101 commented:
@Shurtu-gal @Amzani Ready for review. Hopefull for the last review cycle :)

peter-rr commented:
/update

AayushSaini101 commented on test/integration/validate.test.ts:
```diff
@@ -245,4 +245,23 @@ describe('validate', () => {
         done();
       });
   });
+  // The score of the asyncapi document is 50
```
Removed thanks @peter-rr

peter-rr reviewed (approved):
LGTM 🚀

Amzani reviewed (approved):
LGTM 🚀

Shurtu-gal commented:
/u

Shurtu-gal reviewed (approved):
LGTM

Shurtu-gal commented:
/rtm
